### PR TITLE
Revert "Revert "Redefine scheduler to set learning rate using recursive formula" #14010 (#21463)" and adding legacy epoch WIP

### PR DESCRIPTION
**Pull request by vincentqb**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21711 Revert "Revert "Redefine scheduler to set learning rate using recursive formula" #14010 (#21463)" and adding legacy epoch**

This reverts commit 3889855a5baf77cbdad60b94bcad77fdd7f1b145.

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -190,14 +201,21 @@ def __init__(self, optimizer, step_size, gamma=0.1, last_epoch=-1):
         super(StepLR, self).__init__(optimizer, last_epoch)
 
     def get_lr(self):
+        if (self.last_epoch == 0) or (self.last_epoch % self.step_size != 0):
+            return [group['lr'] for group in self.optimizer.param_groups]
+        return [group['lr'] * self.gamma
+                for group in self.optimizer.param_groups]
+
+    def legacy_get_lr(self):
```
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: IMO, the name `legacy_get_lr` is pretty unclear.  It would be better to name it something that makes it clearer that it computes a closed form based on `last_epoch` and the `base_lrs`.  Maybe also consider taking `last_epoch` explicitly as an argument.

File: test/test_optim.py [hunk 2]
```diff
@@ -1143,7 +1247,7 @@ def _test(self, schedulers, targets, epochs=10):
         if isinstance(schedulers, _LRScheduler):
             schedulers = [schedulers]
         for epoch in range(epochs):
-            [scheduler.step(epoch) for scheduler in schedulers]
+            [scheduler.step() for scheduler in schedulers]
```
Comment (ezyang) [inline test/test_optim.py]: If `epoch` is one greater than `last_epoch` I think we should accept that in the API without complaining, and do the new behavior.

**Comment by ezyang** [comment]
I didn't do a full review but two things popped out to me.

**Comment by vincentqb** [comment]
New PR #21800
